core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -353,41 +350,31 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     */
   def concurrently[F2[x] >: F[x], O2](that: Stream[F2, O2])(
       implicit F: Concurrent[F2]): Stream[F2, O] =
-    Stream.eval(Deferred[F2, Unit]).flatMap { interruptR =>
-      Stream.eval(Deferred[F2, Option[Throwable]]).flatMap { doneR =>
-        Stream.eval(Deferred[F2, Unit]).flatMap { interruptL =>
-          def runR =
-            that
-              .interruptWhen(interruptR.get.attempt)
-              .compile
-              .drain
-              .attempt
-              .map { _.left.toOption }
-              .flatMap { r =>
-                // to prevent deadlock, done must be signalled before `interruptL`
-                // in case the interruptL is signaled before the `L` stream may be in
-                // its `append` code, that requires `get` to complete, which won't ever complete,
-                // b/c it will be evaluated after `interruptL`
-                doneR.complete(r) >>
-                  r.fold(F.unit)(_ => interruptL.complete(()))
+    Stream.eval {
+      Deferred[F2, Unit].flatMap { interrupt =>
+        Deferred[F2, Either[Throwable, Unit]].map { doneR =>
+          def runR: F2[Unit] =
+            that.interruptWhen(interrupt.get.attempt).compile.drain.attempt.flatMap { r =>
```

Comment:
Instread of `.drain.attempt.flatMap`, should this be `.drain.guaranteeCase`? If the `F[_]` is canceled at effect level (not stream level), you still want to run the body of the `flatMap`, right?

Reply:
the `F` cannot be cancelled here. Note the user does not have access to `F` (due to `F.start(runR)`) construct. 

I think whole cancellation story doesn't really play well in fs2, hence the finalisers may not kick in if you cancel. I think we have to make clear distinction between interruption (what we do in fs2, that is also recoverable) and cancellation, that is on `F` and is not recoverable in stream scope. 

It sad, I was hoping we would be able to use cancellation instead of interruption in merge, but that seems to be not possible, or at least I don't know how.

Reply:
> I think whole cancellation story doesn't really play well in fs2, hence the finalisers may not kick in if you cancel. I think we have to make clear distinction between interruption (what we do in fs2, that is also recoverable) and cancellation, that is on F and is not recoverable in stream scope.

Just to be clear, cancelling a compiled stream still guarantees all resources are finalized correctly.

Reply:
Well, I am not sure of that. I think cancellation may or may not run the finalisers. 

What do you think proves that?

Reply:
Ok I take it back. In Algebra#compile there is bracket, so yes, indeed the root scope will get finalised always even when `F` is cancelled, and now, thanks to new `cancel` behaviour user may decide to wait/not on finalisers to complete. 

 I may take another shot after this PR to further optimize `merge` and `concurrently` with this.